lib/lpc55-rot-startup/src/images.rs

```diff
@@ -56,129 +38,439 @@ extern "C" {
 const PAGE_SIZE: u32 = BYTES_PER_FLASH_PAGE as u32;
 
 pub struct Image {
+    // The boundaries of the flash slot.
     flash: Range<u32>,
-    vector: &'static ImageVectors,
+    // The runtime address range (stage0next flash != run).
+    run: Range<u32>,
+    // The contiguous span of programmed flash pages starting at offset zero.
+    // Note that any additional programmed pages after the first erased
+    // page are not interesting for image sanity checks and are not included.
+    programmed: Range<u32>,
+    // Measurement over all pages includeing those that follow any erased page.
+    fwid: [u8; 32],
+    status: Result<(), ImageError>,
 }
 
-pub fn image_details(img: Image, flash: &mut Flash) -> RotImageDetails {
-    RotImageDetails {
-        digest: img.get_fwid(flash),
-        version: img.get_image_version(),
+impl Image {
+    pub fn get_image_b(flash: &mut Flash) -> Image {
+        let mut img = Image::new(flash, FLASH_B, FLASH_B);
+        img.status = img.validate(true);
+        img
+    }
+
+    pub fn get_image_a(flash: &mut Flash) -> Image {
+        let mut img = Image::new(flash, FLASH_A, FLASH_A);
+        img.status = img.validate(true);
+        img
+    }
+
+    pub fn get_image_stage0(flash: &mut Flash) -> Image {
+        let mut img = Image::new(flash, FLASH_STAGE0, FLASH_STAGE0);
+        img.status = img.validate(false);
+        img
+    }
+
+    pub fn get_image_stage0next(flash: &mut Flash) -> Image {
+        let mut img = Image::new(flash, FLASH_STAGE0_NEXT, FLASH_STAGE0);
+        img.status = img.validate(false);
+        img
+    }
```

Comment:
The change was motivated by the need for a fwid measurement on any flash slot. I should have gone further
to keep the caller out of the safety business.

I had thought of splitting Image into FlashSlot and Image where Image could only be constructed if there
was a valid image but we retain the FWID property of the slot. The API messages are changing in any case to
accommodate the ImageError and there always being an FWID.

FWID has changed definition: "...Expanding our
    definition for the RoT FWID as not only the expected image, but also any
    other programmed flash pages in the range of possible pages mitigates a
    potential persistence mechanism for an attacker."

Reply:
I factored out FlashSlot from Image so that fwid is part of the slot and not the image. That allows Image::new() to return Result<Image, ImageError> and avoid having the caller be responsible for safety.